### DISCUSSION TITLE
Query string

### DISCUSSION THREAD

**Discussion by MSIborisyeltsin** [community/community#62300] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

In my opinion, being able to add a query string to the github.io URL would be a cool feature.

## What is the problem?
If I want to add the query string `<link>?attr1=val1&attr2=val2` then instead of the expected result I get a slash between `<link>` and `?`.

## What is the usefulness of the function?
Each user will receive a unique link to the same form. In my case it will help me in creating new readme.md decoration plugins.
To understand what I mean, you can follow the repository of my site on github at [this link](https://github.com/Joomipoll/progressbar)

## P.S.

Sorry for my english :). I'm from Serbia

**Comment by h7x4** [comment]
Hello!

It's unclear to me what you are trying to query. Are you querying repositories? Issues or PRs? Maybe you could give some examples of what `attr1` and `attr2` could be in a real life scenario, and what you think they should return?

**Reply by MSIborisyeltsin** [reply]
> Maybe you could give some examples of what attr1 and attr2 could be in a real life scenario

I mean send GET requests using links. In my case: `https://joomipoll.github.io/progressbar?percent=30%&bgfinish=white` - sets the loading bar percentage and changes the color of the strip gradually to white <br>
But when I enter this link, I get "/" and it breaks: `https://joomipoll.github.io/progressbar/?percent=30%&bgfinish=white`. So, GET requests don't work

For more understanding please read [this article](https://en.wikipedia.org/wiki/HTTP#Request_methods)

**Comment by h7x4** [comment]
Automatically appending `/` is default behaviour of a lot of webservers. When a webserver does not find a page or rule defined at the specified url, it may try to detect if it matches the name of a directory. And when the part behind is omitted, it (usually) implicitly means `index.html`. For example, have a look at [how nginx does it](https://www.nginx.com/resources/wiki/start/topics/tutorials/config_pitfalls/#check-if-file-exists).

This means that the "real" url to your website is actually https://joomipoll.github.io/progressbar/index.html?percent=30%&bgfinish=white. So the slash is not the part that breaks your app.

If you open your javascript console with <kbd>F12</kbd>, you can see that you have access to the query parameters using `window.location.search`

![image](https://github.com/community/community/assets/14929991/b232ece6-75a2-406e-a150-1296d182ef24)

I believe your problem might have something to do with the `</html>` tag you in your `<script>`:

![image](https://github.com/community/community/assets/14929991/e8bcbfe1-d4ff-4e9e-ac06-bcc73bae8cf9)

Maybe you've forgotten to put a `</script>` behind your `<script>` tag?

**Reply by h7x4** [reply]
I just had a look at https://github.com/Joomipoll/progressbar/blob/main/index.html

`<script>` is not a [void tag](https://developer.mozilla.org/en-US/docs/Glossary/Void_element) :)

> If a trailing / (slash) character is present in the start tag of an HTML element, HTML parsers ignore that slash character. This is important to remember when an element such as [`<script>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script) or [`<ul>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul) does require a closing tag. In this case, adding a trailing slash in the start tag does not close the element.

**Reply by MSIborisyeltsin** [reply]
I tried to act differently:

```js
const progressBar = document.getElementById("progressbar");

var queryString = window.location.search;
var params = new URLSearchParams(queryString);

const setStyleProperty = (property, value) => {
  if (value != null)
  {
	document.documentElement.style.setProperty(property, value);
  }
};

const progressBar0 = document.querySelector('#progressbar');

setStyleProperty('--duration', params.get('duration'));
setStyleProperty('--bgclr', params.get('bgclr'));
setStyleProperty('--bgstart', params.get('bgstart'));
setStyleProperty('--bgfinish', params.get('bgfinish'));

if (getComputedStyle(document.documentElement).getPropertyValue('--bgfinish') === 'white')
{
	progressBar0.style.color = "black";
}

const percent = params.get('percent');
if (percent != null)
{
	progressBar0.setAttribute('data-percent', percent);
}
```
But it doesn't work at all
The strangest thing about this is that this code works in the developer console in the browser

To work I used the `https://joomipoll.github.io/progressbar/index.html?percent=30%&bgfinish=white` link+

**Comment by h7x4** [comment]
index.html:

```diff
<!DOCTYPE html>
<html lang="ru">
    <head>
        <meta charset="UTF-8">
        <link rel="stylesheet" href="style.css">
        <title>Test</title>
    </head>
    <body>
        <div id="progressbar" data-percent="1%" style="--width: 1"></div>
-        <script src="script.js"/>
+        <script src="script.js"></script>
    </body>
</html>
```

**Reply by MSIborisyeltsin** [reply]
- if it were an incorrect line, then the animation of the appearance of the loading bar would not work 
- This did not significantly affect the problem in any way

upd: ok i did it

**Reply by h7x4** [reply, marked as answer]
Have a go at merging it into main, to have a look at the difference.

> The strangest thing about this is that this code works in the developer console in the browser

The code in your `script.js` does not run because the `<script>` tag was not closed. Browsers just convert `<script/>` -> `<script>` implicitly. The browser autocompleted your HTML code incorrectly, as shown in this picture:

![](https://user-images.githubusercontent.com/14929991/257762933-e8bcbfe1-d4ff-4e9e-ac06-bcc73bae8cf9.png)

`<script/>` is not valid HTML, the only elements you can use this syntax for are listed here. They are called void elements: https://developer.mozilla.org/en-US/docs/Glossary/Void_element

> then the animation of the appearance of the loading bar would not work

This is handled by the CSS `@keyframes`, it would work even if you deleted `script.js`

**Reply by MSIborisyeltsin** [reply]
Thank you very much!
